baseapp/baseapp.go

```diff
@@ -678,6 +683,8 @@ func (app *BaseApp) runTx(mode runTxMode, txBytes []byte) (gInfo sdk.GasInfo, re
 	// Result if any single message fails or does not have a registered Handler.
 	result, err = app.runMsgs(runMsgCtx, msgs, mode)
 	if err == nil && mode == runTxModeDeliver {
+		consumeBlockGas()
```

Comment:
This is the main fix -- gas is consumed before calling `msCache.Write()`, ensuring a panic is thrown when if the transaction consumes more than the block limit